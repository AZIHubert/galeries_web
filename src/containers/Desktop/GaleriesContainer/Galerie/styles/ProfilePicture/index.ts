import styled from 'styled-components';

const ProfilePicture = styled.div`
  height: 22px;
  width: 22px;
  border: ${({ theme }) => (
    `2px solid ${theme.colors.white}`
  )};
  border-radius: 50%;
  overflow: hidden;
  margin-right: -6px;
`;

export default ProfilePicture;
