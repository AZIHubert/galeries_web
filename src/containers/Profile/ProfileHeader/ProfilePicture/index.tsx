import * as React from 'react';
import { useSelector } from 'react-redux';
import BounceLoader from 'react-spinners/BounceLoader';
import { CSSTransition } from 'react-transition-group';

import theme from '#helpers/theme';

import Image from '#components/Image';

import { ProfilePictureContext } from '#contexts/ProfilePictureContext';

import { userSelector } from '#store/selectors';

import {
  Container,
  Fader,
  ImageContainer,
  InnerContainer,
  SpinnerContainer,
} from './styles';

const ProfilePicture = () => {
  const {
    isPosting,
    deletingImage,
    puttingImage,
    profilePicture,
  } = React.useContext(ProfilePictureContext);

  const user = useSelector(userSelector);

  const isDeletingCurrentProfilePicture = () => {
    if (user && !!deletingImage && user.currentProfilePictureId === deletingImage) {
      return true;
    }
    return false;
  };

  return (
    <Container>
      <InnerContainer>
        <ImageContainer
          isPending={isPosting || !!puttingImage || isDeletingCurrentProfilePicture()}
        >
          <Image
            alt='current profile picture'
            mode='fill'
            original={profilePicture.croped}
            pending={profilePicture.pending}
          />
        </ImageContainer>
        <CSSTransition
          classNames='fade'
          in={isPosting || !!puttingImage || isDeletingCurrentProfilePicture()}
          timeout={300}
          unmountOnExit
        >
          <Fader>
            <SpinnerContainer>
              <BounceLoader
                color={theme.colors.secondary}
                size={60}
              />
            </SpinnerContainer>
          </Fader>
        </CSSTransition>
      </InnerContainer>
    </Container>
  );
};

export default ProfilePicture;
