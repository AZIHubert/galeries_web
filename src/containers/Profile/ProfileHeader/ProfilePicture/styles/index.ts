export { default as Container } from './Container';
export { default as Fader } from './Fader';
export { default as ImageContainer } from './ImageContainer';
export { default as InnerContainer } from './InnerContainer';
export { default as SpinnerContainer } from './SpinnerContainer';
