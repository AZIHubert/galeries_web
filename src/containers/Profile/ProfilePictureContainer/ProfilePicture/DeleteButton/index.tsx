import * as React from 'react';
import { AiOutlineDelete } from 'react-icons/ai';
import { useDispatch } from 'react-redux';
import styled from 'styled-components';

import Modal from '#components/Modal';
import Text from '#components/Text';
import ButtonLol from '#components/Button';

import { ProfilePictureContext } from '#contexts/ProfilePictureContext';

import theme from '#helpers/theme';

import { deleteProfilePicture } from '#store/actions';

import {
  Button,
} from './styles';

interface DeleteButtonI {
  id: string;
}

const ButtonContainer = styled.div`
  display: flex;
  justify-content: space-between;
`;

const DeleteButton = ({
  id,
}: DeleteButtonI) => {
  const dispatch = useDispatch();
  const [open, setOpen] = React.useState<boolean>(false);
  const {
    deletingImage,
    puttingImage,
    setDeletingImage,
  } = React.useContext(ProfilePictureContext);

  const handleClick = () => {
    if (!deletingImage && !puttingImage) {
      setDeletingImage(id);
      dispatch(deleteProfilePicture({ id }));
      setOpen(false);
    }
  };
  const handleClose = () => setOpen(false);
  const handleOpen = () => setOpen(true);

  console.log('from button', open);

  return (
    <Button>
      <AiOutlineDelete
        color={theme.colors.secondary}
        onClick={handleOpen}
        size={20}
      />
      <Modal.Portal
        handleClose={handleClose}
        open={open}
      >
        <Modal.Container
          variant='danger'
        >
          <Text
            color='danger'
            styles={{
              fontSize: 1.3,
              marginBottom: 40,
            }}
          >
            Are you sure you want to delete this profile picture?
          </Text>
          <ButtonContainer>
            <ButtonLol.Default
              danger
              title='yes'
              variant='primary'
              onClick={handleClick}
            />
            <ButtonLol.Default
              danger
              title='no'
              variant='secondary'
              onClick={handleClose}
            />
          </ButtonContainer>
        </Modal.Container>
      </Modal.Portal>
    </Button>
  );
};

export default DeleteButton;
