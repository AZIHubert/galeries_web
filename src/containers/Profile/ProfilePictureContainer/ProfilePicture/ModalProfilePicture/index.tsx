import moment from 'moment';
import * as React from 'react';

import { UserContext } from '#contexts/UserContext';

interface ModalProfilPictureI {
  profilePicture: ProfilePictureI;
}

const formatBytes = (a: number, b = 2) => {
  if (a === 0) return '0 Bytes'; const c = b < 0 ? 0 : b;
  const d = Math.floor(Math.log(a) / Math.log(1024));
  return `${parseFloat((a / 1024 ** d).toFixed(c))} ${['Bytes', 'KB', 'MB', 'GB', 'TB', 'PB', 'EB', 'ZB', 'YB'][d]}`;
};

const ModalProfilPicture = ({
  profilePicture,
}: ModalProfilPictureI) => {
  const {
    createdAt,
    originalImage: {
      height,
      signedUrl,
      size,
      width,
    },
  } = profilePicture;
  const { setUser, user } = React.useContext(UserContext);
  const switchCurrent = () => {
    setUser((prevState) => {
      if (prevState) {
        const { currentProfilePicture } = prevState;
        if (currentProfilePicture && currentProfilePicture.id === profilePicture.id) {
          return {
            ...prevState,
            currentProfilePicture: null,
            currentProfilePictureId: null,
          };
        }
        return {
          ...prevState,
          currentProfilePicture: profilePicture,
          currentProfilePictureId: profilePicture.id,
        };
      }
      return null;
    });
  };
  const current = user
    ? user.currentProfilePictureId === profilePicture.id
    : null;
  return (
    <div>
      <img src={signedUrl} alt='image' />
      <p>
        upload at
      </p>
      <p
        data-testid='uploadAtText'
      >
        {` ${moment(createdAt).format('MMMM Do YYYY')}`}
      </p>
      <p>
        weight:
      </p>
      <p
        data-testid='weightText'
      >
        {formatBytes(size)}
      </p>
      <p>
        size:
      </p>
      <p
        data-testid='sizeText'
      >
        {`${width} x ${height} px`}
      </p>
      <button
        data-testid='modalProfilePictureButton'
        onClick={() => switchCurrent()}
      >
        {current ? 'remove profile picture' : 'use as profile picture'}
      </button>
      <button>
        download image
      </button>
    </div>
  );
};

export default ModalProfilPicture;
