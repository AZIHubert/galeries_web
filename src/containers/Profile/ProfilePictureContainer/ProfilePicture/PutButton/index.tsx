import * as React from 'react';
import {
  useDispatch,
  useSelector,
} from 'react-redux';

import { ProfilePictureContext } from '#contexts/ProfilePictureContext';

import { putProfilePicture } from '#store/actions';
import { userSelector } from '#store/selectors';

import {
  Container,
} from './styles';

interface PutButtonI {
  id: string;
}

const PutButton = ({
  id,
}: PutButtonI) => {
  const {
    deletingImage,
    puttingImage,
    setPuttingImage,
  } = React.useContext(ProfilePictureContext);
  const dispatch = useDispatch();
  const user = useSelector(userSelector);

  const isCurrent = () => {
    if (user && user.currentProfilePictureId === id) {
      return true;
    }
    return false;
  };

  const handleClick = () => {
    if (!puttingImage && !deletingImage) {
      dispatch(putProfilePicture({ id }));
      setPuttingImage(id);
    }
  };

  return (
    <Container
      current={isCurrent()}
      onClick={handleClick}
    />
  );
};

export default PutButton;
