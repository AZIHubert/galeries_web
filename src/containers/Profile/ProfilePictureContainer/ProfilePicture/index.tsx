import * as React from 'react';

import { ProfilePictureI } from '#helpers/interfaces';

interface ProfilePictureComponentI {
  profilePicture: ProfilePictureI;
  switchCurrent: () => void
}

const ProfilePicture = ({
  profilePicture: {
    originalImage: {
      signedUrl,
    },
  },
  switchCurrent,
}: ProfilePictureComponentI) => (
  <div>
    <button
      data-testid='profilePictureButton'
      onClick={switchCurrent}
    />
    <a href='#'>
      <img
        src={signedUrl}
        alt='profile picture'
      />
    </a>
  </div>
);

export default ProfilePicture;
