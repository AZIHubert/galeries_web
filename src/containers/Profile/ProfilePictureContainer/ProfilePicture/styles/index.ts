export { default as Container } from './Container';
export { default as Link } from './Link';
export { default as Opacity } from './Opacity';
export { default as Fader } from './Fader';
export { default as SpinnerContainer } from './SpinnerContainer';
