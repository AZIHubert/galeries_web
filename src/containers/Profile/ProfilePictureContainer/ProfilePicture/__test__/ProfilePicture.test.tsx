import * as React from 'react';
import renderer from 'react-test-renderer';
import {
  fireEvent,
  render,
  screen,
} from '@testing-library/react';

import {
  ImageI,
  ProfilePictureI,
} from '#helpers/interfaces';

import ProfilePicture from '../index';

const newImage: ImageI = {
  bucketName: 'bucketName',
  fileName: 'fileName',
  format: 'jpg',
  height: 2340,
  id: '1',
  signedUrl: 'http://url.com',
  size: 100857,
  width: 1080,
};
const newProfilePicture: ProfilePictureI = {
  createdAt: new Date('2021-02-05 18:42:02.528+01'),
  cropedImage: newImage,
  id: '1',
  originalImage: newImage,
  pendingImage: newImage,
};

const Container = () => {
  const [current, setCurrent] = React.useState<boolean>(false);
  const switchCurrent = () => {
    setCurrent((prevState) => !prevState);
  };
  return (
    <>
      <ProfilePicture
        profilePicture={newProfilePicture}
        switchCurrent={switchCurrent}
      />
      <p
        data-testid='checkCurrentState'
      >
        {current ? 'true' : 'false'}
      </p>
    </>
  );
};

describe('ProfilePicture', () => {
  const mockSwitchCurrent = jest.fn;
  it('renders without crashing', () => {
    const tree = renderer.create(<ProfilePicture
      switchCurrent={mockSwitchCurrent}
      profilePicture={newProfilePicture}
    />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('shouls switch current', () => {
    const { getByTestId } = render(<Container />);
    const profilePictureButton = getByTestId('profilePictureButton');
    fireEvent.click(profilePictureButton);
    const checkCurrentStateTrue = screen.getByTestId('checkCurrentState');
    expect(checkCurrentStateTrue).toHaveTextContent('true');
    fireEvent.click(profilePictureButton);
    const checkCurrentStateFalse = screen.getByTestId('checkCurrentState');
    expect(checkCurrentStateFalse).toHaveTextContent('false');
  });
});
