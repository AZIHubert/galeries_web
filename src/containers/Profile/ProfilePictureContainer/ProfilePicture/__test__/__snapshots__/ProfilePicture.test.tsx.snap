// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePicture renders without crashing 1`] = `
<div>
  <button
    data-testid="profilePictureButton"
    onClick={[Function]}
  />
  <a
    href="#"
  >
    <img
      alt="profile picture"
      src="http://url.com"
    />
  </a>
</div>
`;
